### PR TITLE
use Bayesian bootstrapping to compute weights

### PR DESCRIPTION
**Pull request by aloctavodia**
Use Bayesian bootstrapping to compute Akaike-like weights, following recommendations in this [paper](https://arxiv.org/abs/1704.02030). I will add other features (like stacking) in a future PR.

### CODE REVIEW

File: pymc3/stats.py [hunk 1]
```diff
@@ -349,6 +353,19 @@ def compare(traces, models, ic='WAIC'):
     ic : string
         Information Criterion (WAIC or LOO) used to compare models.
         Default WAIC.
+    bootstrapping : boolean
```
Comment (aseyboldt) [inline pymc3/stats.py]: I think I'd prefer `bootstrap`

File: pymc3/stats.py [hunk 2]
```diff
@@ -405,16 +423,43 @@ def add_warns(*args):
...
+
+        b_weighting = dirichlet.rvs(alpha=[alpha] * N, size=b_samples,
+                                    random_state=seed)
+        weights = np.zeros((b_samples, len(ics)))
+        z_bs = np.zeros((b_samples, len(ics)))
+        for i in range(b_samples):
+            z_b = np.dot(ic_i, b_weighting[i])
+            u_weights = np.exp(-0.5 * (z_b - np.min(z_b)))
+            z_bs[i] = z_b
+            weights[i] = u_weights / np.sum(u_weights)
```
Comment (aseyboldt) [inline pymc3/stats.py:441]: This is pretty much `softmax`, right? That doesn't seem to be particularly stable that way, something like [this](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/extmath.py#L584) might be better.

File: pymc3/stats.py (hunk 1 above)
Comment (aloctavodia) [reply pymc3/stats.py]: OK, I will change it

File: pymc3/stats.py (hunk 2 above)
Comment (aloctavodia) [reply pymc3/stats.py:441]: Yes, this is basically softmax and this is already an stable version of it (notice that I am subtracting the min value of z_b)

**Comment by aseyboldt** [comment]
You are right, sorry.

On August 5, 2017 2:41:50 PM CEST, Osvaldo Martin <notifications@github.com> wrote:
>aloctavodia commented on this pull request.
>
>
>
>> +    if bootstrapping:
>+        N = len(ics[0][1][3])
>+
>+        ic_i = np.zeros((len(ics), N))
>+        for i in range(len(ics)):
>+            ic_i[i] = ics[i][1][3] * N
>+
>+        b_weighting = dirichlet.rvs(alpha=[alpha] * N, size=b_samples,
>+                                    random_state=seed)
>+        weights = np.zeros((b_samples, len(ics)))
>+        z_bs = np.zeros((b_samples, len(ics)))
>+        for i in range(b_samples):
>+            z_b = np.dot(ic_i, b_weighting[i])
>+            u_weights = np.exp(-0.5 * (z_b - np.min(z_b)))
>+            z_bs[i] = z_b
>+            weights[i] = u_weights / np.sum(u_weights)
>
>Yes, this is basically softmax and this is already an stable version of
>it (notice that I am subtracting the min value of z_b)
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pymc-devs/pymc3/pull/2479#discussion_r131521398

-Adrian

**Comment by aseyboldt** [comment]
I see, you are right of course.

On August 5, 2017 2:41:50 PM CEST, Osvaldo Martin <notifications@github.com> wrote:
>aloctavodia commented on this pull request.
>
>
>
>> +    if bootstrapping:
>+        N = len(ics[0][1][3])
>+
>+        ic_i = np.zeros((len(ics), N))
>+        for i in range(len(ics)):
>+            ic_i[i] = ics[i][1][3] * N
>+
>+        b_weighting = dirichlet.rvs(alpha=[alpha] * N, size=b_samples,
>+                                    random_state=seed)
>+        weights = np.zeros((b_samples, len(ics)))
>+        z_bs = np.zeros((b_samples, len(ics)))
>+        for i in range(b_samples):
>+            z_b = np.dot(ic_i, b_weighting[i])
>+            u_weights = np.exp(-0.5 * (z_b - np.min(z_b)))
>+            z_bs[i] = z_b
>+            weights[i] = u_weights / np.sum(u_weights)
>
>Yes, this is basically softmax and this is already an stable version of
>it (notice that I am subtracting the min value of z_b)
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pymc-devs/pymc3/pull/2479#discussion_r131521398

-Adrian

**Comment by aloctavodia** [comment]
Thanks @aseyboldt for your review and for pointing out how to fix the slow-down problem! I think this is ready to merge, right?

**Comment by junpenglao** [comment]
Yep LGTM ;-)
